CMakeLists.txt

```diff
@@ -153,93 +153,68 @@ endif()
-toolchain_cc_cpp_dialect_std_17_flags(CPP_DIALECT_STD_17_FLAGS)
-toolchain_cc_cpp_dialect_std_2a_flags(CPP_DIALECT_STD_2A_FLAGS)
-
 if(CONFIG_CPLUSPLUS)
   # From kconfig choice, pick a single dialect.
   # Kconfig choice ensures only one of these CONFIG_STD_CPP* is set.
   if(CONFIG_STD_CPP98)
-    set(STD_CPP_DIALECT_FLAGS ${CPP_DIALECT_STD_98_FLAGS})
+    set(STD_CPP_DIALECT_FLAGS $<TARGET_PROPERTY:compiler-cpp,dialect_cpp98>)
   elseif(CONFIG_STD_CPP11)
-    set(STD_CPP_DIALECT_FLAGS ${CPP_DIALECT_STD_11_FLAGS}) # Default in kconfig
+    set(STD_CPP_DIALECT_FLAGS $<TARGET_PROPERTY:compiler-cpp,dialect_cpp11>) # Default in kconfig
   elseif(CONFIG_STD_CPP14)
-    set(STD_CPP_DIALECT_FLAGS ${CPP_DIALECT_STD_14_FLAGS})
+    set(STD_CPP_DIALECT_FLAGS $<TARGET_PROPERTY:compiler-cpp,dialect_cpp14>)
   elseif(CONFIG_STD_CPP17)
-    set(STD_CPP_DIALECT_FLAGS ${CPP_DIALECT_STD_17_FLAGS})
+    set(STD_CPP_DIALECT_FLAGS $<TARGET_PROPERTY:compiler-cpp,dialect_cpp17>)
   elseif(CONFIG_STD_CPP2A)
-    set(STD_CPP_DIALECT_FLAGS ${CPP_DIALECT_STD_2A_FLAGS})
+    set(STD_CPP_DIALECT_FLAGS $<TARGET_PROPERTY:compiler-cpp,dialect_cpp2a>)
   else()
     assert(0 "Unreachable code. Expected C++ standard to have been chosen. See Kconfig.zephyr.")
   endif()
 
-  foreach(flag ${STD_CPP_DIALECT_FLAGS})
-    zephyr_compile_options(
-      $<$<COMPILE_LANGUAGE:CXX>:${flag}>
-    )
-  endforeach()
+  zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:${STD_CPP_DIALECT_FLAGS}>)
 endif()
 
 if(NOT CONFIG_EXCEPTIONS)
   # @Intent: Obtain compiler specific flags related to C++ Exceptions
-  toolchain_cc_cpp_no_exceptions_flag(CPP_NO_EXCEPTIONS_FLAG)
-  zephyr_compile_options(
-    $<$<COMPILE_LANGUAGE:CXX>:${CPP_NO_EXCEPTIONS_FLAG}>
-  )
+  zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:$<TARGET_PROPERTY:compiler-cpp,no_exceptions>>)
 endif()
 
 if(NOT CONFIG_RTTI)
   # @Intent: Obtain compiler specific flags related to C++ Run Time Type Information
-  toolchain_cc_cpp_no_rtti_flag(CPP_NO_RTTI_FLAG)
-  zephyr_compile_options(
-    $<$<COMPILE_LANGUAGE:CXX>:${CPP_NO_RTTI_FLAG}>
-  )
+  zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:$<TARGET_PROPERTY:compiler-cpp,no_rtti>>)
 endif()
 
 if(CONFIG_MISRA_SANE)
   # @Intent: Obtain toolchain compiler flags relating to MISRA.
-  toolchain_cc_warning_error_misra_sane(CC_MISRA_SANE_FLAG)
-  toolchain_cc_cpp_warning_error_misra_sane(CPP_MISRA_SANE_FLAG)
-  zephyr_compile_options($<$<COMPILE_LANGUAGE:C>:${CC_MISRA_SANE_FLAG}>)
-  zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:${CPP_MISRA_SANE_FLAG}>)
+  zephyr_compile_options($<$<COMPILE_LANGUAGE:C>:$<TARGET_PROPERTY:compiler,warning_error_misra_sane>>)
+  zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:$<TARGET_PROPERTY:compiler-cpp,warning_error_misra_sane>>)
 endif()
 
 # @Intent: Set compiler specific macro inclusion of AUTOCONF_H
-toolchain_cc_imacros(${AUTOCONF_H})
+zephyr_compile_options($<TARGET_PROPERTY:compiler,imacros>${AUTOCONF_H})
```

Comment:
Need the ability to add a space, in our case, or some other assignment character between the imacros property and the string result from ${AUTOCONF_H}.
Setting the imacros property as follows, seems to work in our case, but the resulting option gets enclosed in " which is probably not desirable.
`set_compiler_property(PROPERTY imacros "-imacros ")`

Reply:
thanks for letting me know.
This is exactly why it's still a draft, in order to collect all the knowledge needed to make this abstraction as useful as possible.

Will try to include in next update round.

Reply:
As agreed at the last Zephyr Toolchain Working Group meeting on May 14th, I have looked into our Toolchain/compiler support for having a single -imacros include file. This file should then #include both the autoconf.h and the zephyr_stdint.h files, thus needing only one -imacros compiler option and header.
The test was performed with the current macro/function based toolchain abstraction, and thus not with the property based toolchain abstraction from this PR, but that makes no difference in this case.

The  toolchain_cc_imacros was modified to look as follows:
```
macro(toolchain_cc_imacros header_file)
  # Generate the common imacros include file, together with a verification header
  #  which simply produces a warning every time the file is included...
  if(NOT ${header_file} MATCHES "/zephyr_.*\.h") \
    file(WRITE /tmp/z_imacros_verf.h "#warning \"-imacros include z_imacros_verf.h\"\n")
    file(WRITE /tmp/z_imacros.h "/* Auto generated, do not edit */\n\n"
                                "#include </tmp/z_imacros_verf.h>\n")
    # Add the -imacros option to the compiler flags
    zephyr_compile_options(-imacros /tmp/z_imacros.h)
  endif()
  # Append the #include of the given header file
  file(APPEND /tmp/z_imacros.h "#include <${header_file}>\n")
endmacro()
```
The above results in a generated /tmp/z_imacros.h which #include autoconf.h and zephyr_stdint.h. 
For our compiler, **this works fine**, it spits out the warning for the z_imacros_verf.h together with including the information needed from the autoconf.h and zephyr_stdint.h header files.
If other compilers supports this as well, it could be that a common imacros header should be generated, similar to above macro, but just more generic. If it turns out that not all compilers support the #include in the -imacros include file, then perhaps there should exist a way a given toolchain could have such a header file generated and added as compile option, similar to the flexibility of the toolchain_cc_imacros macro.

On a side note, I actually discovered that the reason our compiler does not accept -imacros inclusion of the two header files, is not that it does not accept two -imacros options, but rather that the zephyr_compile_options, hence target_compile_options, strips one of the -imacros options leaving only one -imacros option with two header files separated by spaces, and that the compiler does not like. The zephyr_compile_options strips the last -imacros option, as it sees it as a duplicate, since we have to have a space between the -imacros and the header file.
Adding the two -imacros options as `zephyr_compile_options("-include ${header_file}")` actually works for our compiler, but it adds " around each -imacros option.

Reply:
> Need the ability to add a space, in our case, or some other assignment character between the imacros property and the string result from ${AUTOCONF_H}.

Is  this a compiler-specific requirement? Just making sure.

> the reason our compiler does not accept -imacros inclusion of the two header files, is not that it does not accept two -imacros options, but rather that the zephyr_compile_options, hence target_compile_options, strips one of the -imacros options leaving only one -imacros option with two header files separated by spaces,

I understood CMake performs flag deduplication, correct? I mean not related and not specific to Zephyr. Not specific to `-imacros` either.

That's why there's no space, as a workaround: https://github.com/zephyrproject-rtos/zephyr/issues/7216#issuecomment-646433554

Reply:
> Is this a compiler-specific requirement? Just making sure.

Yes, it is a compiler specific requirement. In our case, the compiler only accepts a space between the option and the header file name, nothing else.

>  I understood CMake performs flag deduplication, correct? I mean not related and not specific to Zephyr. Not specific to -imacros either. 

That is true, CMake performs flag deduplication on any duplicated flags, not specific to -imacros, nor to Zephyr. Perhaps @tejlmand could confirm that?

>  That's why there's no space, as a workaround: #7216 (comment)

Yes, but as stated previously, not all compilers support the 'no space' workaround. Which is why we need a way to circumvent the CMake flag deduplication, and also handle those cases where the compiler only supports one -imacros option.

Reply:
> > Is this a compiler-specific requirement? Just making sure.
> 
> Yes, it is a compiler specific requirement. In our case, the compiler only accepts a space between the option and the header file name, nothing else.
> 
> > I understood CMake performs flag deduplication, correct? I mean not related and not specific to Zephyr. Not specific to -imacros either.
> 
> That is true, CMake performs flag deduplication on any duplicated flags, not specific to -imacros, nor to Zephyr. Perhaps @tejlmand could confirm that?
> 

Well, that was true for older CMake versions, but since CMake v3.12, you can avoid the deduplication:
>While beneficial for individual options, the de-duplication step can break up option groups. For example, -D A -D B becomes -D A B. One may specify a group of options using shell-like quoting along with a SHELL: prefix. The SHELL: prefix is dropped and the rest of the option string is parsed using the separate_arguments() UNIX_COMMAND mode. For example, "SHELL:-D A" "SHELL:-D B" becomes -D A -D B.

https://cmake.org/cmake/help/v3.12/command/target_compile_options.html

Reply:
finally had time to try it out in this PR, and here is the result when compiling `main.c`:
```
/opt/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc                                      
-DBUILD_VERSION=zephyr-v2.3.0-969-g2811e4ecdde1                                                     
-DKERNEL                                                                                            
-DNRF52840_XXAA                                                                                     
-D_FORTIFY_SOURCE=2                                                                                 
-D__PROGRAM_START                                                                                   
-D__ZEPHYR__=1                                                                                      
...
-isystem /projects/github/ncs/zephyr/lib/libc/minimal/include                                                
-isystem /opt/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include                 
-isystem /opt/zephyr-sdk-0.11.3/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include-fixed           
-Os
-imacros /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/include/generated/autoconf.h           
-ffreestanding                                                                                      
-fno-common                                                                                         
-g                                                                                                  
-mcpu=cortex-m4                                                                                     
-mthumb                                                                                             
-mabi=aapcs                                                                                         
-imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h                                       
...
-o CMakeFiles/app.dir/src/main.c.obj                                                                   
-c ../src/main.c                                      
```

and as it can be seen, there is not deduplications of `-imacro`.
And there is even a `space` between the `-imacro` and the header file.
@daor-oti Going to push an update soon, so that it can be tested with custom compilers.

Reply:
@daor-oti feel free to try the updated version with your custom compiler.

Reply:
@tejlmand Thank you, this looks good and is working perfectly with our compiler, hence no deduplication of -imacros. Also the, in our case, needed space is there without us having to explicitly add the space to the imacros property, nice.